fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/impl/DefaultOCFLObjectSession.java

```diff
@@ -89,85 +115,267 @@ public void write(final String subpath, final InputStream stream) {
+        return !ocflRepository.getObject(ObjectVersionId.head(objectIdentifier))
+                .containsFile(subpath);
     }
 
     /**
      * {@inheritDoc}
      */
     @Override
-    public InputStream read(final String subpath) {
-        if (hasStagedChanges(subpath)) {
-            // TODO read the staged version of the file
-        } else {
-            // TODO read the head version of the file from the ocfl object
+    public InputStream read(final String subpath) throws PersistentItemNotFoundException {
+        final var stagedPath = resolveStagedPath(subpath);
+
+        if (hasStagedChanges(stagedPath)) {
+            // prioritize read of the staged version of the file
+            try {
+                return new FileInputStream(stagedPath.toFile());
+            } catch (final FileNotFoundException e) {
+                throw new PersistentItemNotFoundException(format("Could not find %s within object %s",
+                        subpath, objectIdentifier));
+            }
+        } else if (!objectDeleted) {
+            // Fall back to the head version
+            return readVersion(subpath, ObjectVersionId.head(objectIdentifier));
         }
-        return null;
-    }
 
-    private boolean hasStagedChanges(final String subpath) {
-        return false;
+        throw new PersistentItemNotFoundException(format("Could not find %s within object %s",
+                subpath, objectIdentifier));
     }
 
     /**
      * {@inheritDoc}
      */
     @Override
-    public InputStream read(final String subpath, final String version) {
-        final AtomicReference<InputStream> contentRef = new AtomicReference<>();
+    public InputStream read(final String subpath, final String version) throws PersistentItemNotFoundException {
+        // If the object was deleted, then only uncommitted staged files can be available
+        if (objectDeleted) {
+            throw new PersistentItemNotFoundException(format("Could not find %s within object %s",
+                    subpath, objectIdentifier));
+        }
 
-        ocflRepository.readObject(ObjectVersionId.version(objectIdentifier, version), reader -> {
-            contentRef.set(reader.getFile(subpath));
-        });
+        return readVersion(subpath, ObjectVersionId.version(objectIdentifier, version));
+    }
 
-        return contentRef.get();
+    private InputStream readVersion(final String subpath, final ObjectVersionId version) throws PersistentItemNotFoundException {
+        try {
+            // read the head version of the file from the ocfl object
+            final var file = ocflRepository.getObject(version)
+                    .getFile(subpath);
+            if (file == null) {
+                throw new PersistentItemNotFoundException(format("Could not find %s within object %s version %s",
+                        subpath, objectIdentifier, version.getVersionId()));
+            }
+            return file.getStream();
```

Comment:
I will disable it for the time being, but that is potentially valuable to know in case its useful for performing fixity checks later. I am guessing it uses whatever algorithm's digest is in the inventory? I would suggest that the `on()` method return the stream rather than void though

Reply:
Yeah, I can do that. I'll add a new method and let you know when it's there. `on` is a method from the underlying MessageDigestInputStream.